### PR TITLE
fix(python): raise error when pandas df's some index name duplicates some column name

### LINKED ISSUE #15938: `pl.from_pandas(df, include_index=True)` silently ignores index if it is named the same way as a column
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
df = pd.DataFrame({'a': [1,2,3], 'b': [4,5,6]}, index=pd.Index([7,8,9], name='a'))
print(pl.from_pandas(df, include_index=True))
```

### Log output

```shell
shape: (3, 2)
┌─────┬─────┐
│ a   ┆ b   │
│ --- ┆ --- │
│ i64 ┆ i64 │
╞═════╪═════╡
│ 1   ┆ 4   │
│ 2   ┆ 5   │
│ 3   ┆ 6   │
└─────┴─────┘
```

### Issue description

the index was silently excluded

### Expected behavior

I think this should raise

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.10
Index type:           UInt32
Platform:             Linux-5.15.146.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
Python:               3.11.9 (main, Apr  6 2024, 17:59:24) [GCC 11.4.0]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          3.0.0
connectorx:           <not installed>
deltalake:            <not installed>
fsspec:               2024.3.1
gevent:               <not installed>
hvplot:               0.9.2
matplotlib:           3.8.4
numpy:                1.26.4
openpyxl:             <not installed>
pandas:               2.2.2
pyarrow:              15.0.2
pydantic:             <not installed>
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           <not installed>
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>
```

</details>

### PR DESCRIPTION
**Pull request by piri-p**
Fix #15938 : raise ValueError when converting from pandas DF, whose index name duplicates some column name @MarcoGorelli

### CODE REVIEW

File: py-polars/polars/_utils/construction/dataframe.py [hunk 1]
```diff
@@ -1020,6 +1020,9 @@ def pandas_to_pydf(
 ) -> PyDataFrame:
     """Construct a PyDataFrame from a pandas DataFrame."""
     convert_index = include_index and not _pandas_has_default_index(data)
+    if convert_index and (data.index.name in data.columns):
+        msg = "cannot convert pandas DataFrame when index name duplicates some column name"
+        raise ValueError(msg)
```
Comment (mcrumiller) [inline py-polars/polars/_utils/construction/dataframe.py:1025]: This solution fails if the dataframe has a MultiIndex:

```python
import pandas as pd
import polars as pl

df = pd.DataFrame(
    {
        "a": [1, 2, 3],
        "b": [1, 2, 3],
    },
    index=pd.MultiIndex.from_arrays([(1, 2, 3), (1, 2, 3)], names=["a", "b"])
)
pl.from_pandas(df, include_index=True)
# shape: (3, 2)
# ┌─────┬─────┐
# │ a   ┆ b   │
# │ --- ┆ --- │
# │ i64 ┆ i64 │
# ╞═════╪═════╡
# │ 1   ┆ 1   │
# │ 2   ┆ 2   │
# │ 3   ┆ 3   │
# └─────┴─────┘
```
Also, please add unit tests. I would recommend two separate functions, one similar to the one above that checks for overlap in a MultiIndex, and another that checks with a normal index.

File: py-polars/tests/unit/dataframe/test_from_pandas.py
Comment (mcrumiller) [inline py-polars/tests/unit/dataframe/test_from_pandas.py:1]: Also--these tests should go into [tests/unit/interop/test_interop.py](https://github.com/pola-rs/polars/blob/main/py-polars/tests/unit/interop/test_interop.py), which is where the other `from_pandas` tests are.

File: py-polars/polars/_utils/construction/dataframe.py [hunk 2]
```diff
@@ -1020,6 +1020,11 @@ def pandas_to_pydf(
 ) -> PyDataFrame:
     """Construct a PyDataFrame from a pandas DataFrame."""
     convert_index = include_index and not _pandas_has_default_index(data)
+    if convert_index and set(data.index.names).intersection(data.columns):
```
Comment (MarcoGorelli) [inline py-polars/polars/_utils/construction/dataframe.py:1023]: Polars converts non-string column names to string

I think a more generic solution is needed, to handle https://github.com/pola-rs/polars/issues/16025 too

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
thanks for your pr, not sure this is general enough

**Pull request closed without merging**